File: backend/quantailabs_patch/risk/position_sizing.py
from dataclasses import dataclass
from decimal import Decimal, getcontext


getcontext().prec = 28

@dataclass
class PositionSizer:
    base_risk_per_trade_pct: float = 0.5  # % of equity to risk if SL hit
    atr_reference_pct: float = 2.0        # baseline ATR% used for scaling risk
    atr_ceiling_pct: float = 6.0          # clamp to avoid overscaling in extreme volatility

    def compute_size(
        self,
        equity_usd: float,
        entry_price: float,
        stop_price: float,
        atr_pct: float | None = None,
    ) -> float:
        """
        Returns quantity (base units) to buy/sell such that risk at SL ~= scaled risk of equity.
        ATR-based scaling reduces exposure during high volatility and gently increases it during quiet regimes.
        """
        entry = Decimal(str(entry_price))
        stop = Decimal(str(stop_price))
        equity = Decimal(str(equity_usd))
        if entry <= 0 or equity <= 0:
            return 0.0

        stop_distance = abs(entry - stop)
        if stop_distance <= 0:
            return 0.0

        risk_pct = Decimal(str(self.base_risk_per_trade_pct))
        if atr_pct is not None and atr_pct > 0:
            atr_value = Decimal(str(atr_pct))
            baseline = Decimal(str(max(self.atr_reference_pct, 1e-6)))
            ceiling = Decimal(str(max(self.atr_ceiling_pct, self.atr_reference_pct)))
            normalized = min(ceiling, max(atr_value, Decimal('0.0001'))) / baseline
            if normalized > 1:
                scale = Decimal('1') / normalized
                risk_pct *= max(Decimal('0.35'), scale)
            else:
                boost = (Decimal('1') - normalized) * Decimal('0.35')
                risk_pct *= Decimal('1') + boost

        risk_usd = equity * (risk_pct / Decimal('100'))
        if risk_usd <= 0:
            return 0.0

        qty = risk_usd / stop_distance
        return float(max(Decimal('0'), qty))

    @staticmethod
    def r_multiple(entry: float, stop: float, price: float, side: str) -> float:
        risk = abs(entry - stop)
        if risk == 0:
            return 0.0
        if side.lower() == "long":
            return (price - entry) / risk
        else:
            return (entry - price) / risk

File: backend/src/quantai/risk/positionSizing.ts
export type PositionSizingParams = {
  equityUsd: number;
  entryPrice: number;
  stopPrice?: number;
  stopDistanceAbs?: number;
  riskPct?: number;
  qualityMultiplier?: number;
  maxNotionalUsd?: number;
  currentAtrPct?: number;
  targetAtrPct?: number;
  minRiskPct?: number;
  maxRiskPct?: number;
};

export type PositionSizingResult = {
  qty: number;
  notionalUsd: number;
  riskUsd: number;
  riskPct: number;
  stopDistance: number;
  rawQty: number;
  rawNotionalUsd: number;
  qualityMultiplier: number;
};

export class PositionSizer {
  constructor(private readonly baseRiskPerTradePct: number) {}

  computeSize(params: PositionSizingParams): PositionSizingResult {
    const baseRiskPct = params.riskPct && params.riskPct > 0 ? params.riskPct : this.baseRiskPerTradePct;
    let riskPct = baseRiskPct;
    const currentAtr = params.currentAtrPct && params.currentAtrPct > 0 ? params.currentAtrPct : null;
    const targetAtr = params.targetAtrPct && params.targetAtrPct > 0 ? params.targetAtrPct : currentAtr;
    if (currentAtr && targetAtr) {
      const ratio = targetAtr / currentAtr;
      const multiplier = Math.max(0.4, Math.min(1.6, ratio));
      const unclamped = baseRiskPct * multiplier;
      const minRisk = params.minRiskPct != null ? params.minRiskPct : baseRiskPct * 0.4;
      const maxRisk = params.maxRiskPct != null ? params.maxRiskPct : baseRiskPct * 1.6;
      riskPct = Math.max(minRisk, Math.min(maxRisk, unclamped));
    }
    const stopDistance = params.stopDistanceAbs != null && params.stopDistanceAbs > 0
      ? params.stopDistanceAbs
      : Math.abs(params.entryPrice - (params.stopPrice ?? 0));
    const qualityMultiplier = params.qualityMultiplier != null && params.qualityMultiplier > 0
      ? params.qualityMultiplier
      : 1;
    if (!(stopDistance > 0) || !(params.entryPrice > 0) || !(params.equityUsd > 0)) {
      return {
        qty: 0,
        notionalUsd: 0,
        riskUsd: 0,
        riskPct,
        stopDistance,
        rawQty: 0,
        rawNotionalUsd: 0,
        qualityMultiplier,
      };
    }
    const riskUsd = params.equityUsd * (riskPct / 100) * qualityMultiplier;
    const rawQty = riskUsd / stopDistance;
    const rawNotionalUsd = rawQty * params.entryPrice;
    const cappedNotional = params.maxNotionalUsd && params.maxNotionalUsd > 0
      ? Math.min(rawNotionalUsd, params.maxNotionalUsd)
      : rawNotionalUsd;
    const adjustedQty = params.entryPrice > 0 ? cappedNotional / params.entryPrice : 0;
    return {
      qty: Math.max(0, adjustedQty),
      notionalUsd: Math.max(0, cappedNotional),
      riskUsd: Math.max(0, riskUsd),
      riskPct,
      stopDistance,
      rawQty: Math.max(0, rawQty),
      rawNotionalUsd: Math.max(0, rawNotionalUsd),
      qualityMultiplier,
    };
  }

  static rMultiple(entry: number, stop: number, price: number, side: 'long' | 'short'): number {
    const risk = Math.abs(entry - stop);
    if (!(risk > 0)) return 0;
    if (side === 'long') {
      return (price - entry) / risk;
    }
    return (entry - price) / risk;
  }
}
